src/jdk.compiler/share/classes/com/sun/tools/javac/comp/LambdaToMethod.java

```diff
@@ -1439,13 +1434,16 @@ private LambdaTranslationContext analyzeLambda(JCLambda tree, String statKey) {
 
         @Override
         public void visitMethodDef(JCMethodDecl tree) {
+            List<ClassSymbol> prevTypesUnderConstruction = typesUnderConstruction;
             List<Frame> prevStack = frameStack;
             try {
+                if (TreeInfo.hasAnyConstructorCall(tree))   // start early construction context
```

Comment:
Shouldn't we just test if we're in a constructor here?

Reply:
Technically, not all constructors invoke `super()` or `this()`, albeit so far, `java.lang.Object` is the only exception.

Of course, it's unlikely that any day soon there will be other examples of this, or that the `java.lang.Object()` constructor will contain a lambda, but if either of those two things ever does happen then this code will still be correct :)

Reply:
> Technically, not all constructors invoke `super()` or `this()`, albeit so far, `java.lang.Object` is the only exception.
> 
> Of course, it's unlikely that any day soon there will be other examples of this, or that the `java.lang.Object()` constructor will contain a lambda, but if either of those two things ever does happen then this code will still be correct :)

Uhm ok. I asked because scanning for a constructor is a bit heavy. And, now that you bring up Object, perhaps one might argue that the entire constructor body in Object represents a pre-construction context. All things considered, given the remote nature of the scenario you propose (Object is so special that it doesn't really count, and we can even check specifically for it, if needed), combined with the computationally-heavy nature of the method you call, I have a mild preference for just checking if it's a constructor.

Reply:
I agree it's not worth the trouble to scan - fixed in 6aa8a9a183b.

I added a comment to hint at the minor discrepancy.